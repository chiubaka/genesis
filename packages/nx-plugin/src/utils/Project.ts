import {
  getWorkspaceLayout,
  names,
  ProjectType as NxProjectType,
  Tree,
} from "@nrwl/devkit";
import path from "node:path";

export type ProjectType = NxProjectType | "e2e";

export interface ProjectNames {
  camelCase: string;
  kebabCase: string;
  pascalCase: string;
}

export class Project {
  private tree: Tree;
  private names: ProjectNames;
  private type: ProjectType;

  constructor(tree: Tree, name: string, type: ProjectType) {
    this.tree = tree;
    this.type = type;

    const {
      className: pascalCase,
      fileName: kebabCase,
      propertyName: camelCase,
    } = names(name);

    this.names = {
      camelCase,
      kebabCase,
      pascalCase,
    };
  }

  public path(relativePath = "") {
    return this.baseDir(path.join(this.getName(), relativePath));
  }

  public relativePath(relativePath = "") {
    const baseDirName = path.basename(this.baseDir());

    return path.join(baseDirName, this.getName(), relativePath);
  }

  public srcPath(relativePath = "") {
    return path.join(this.path("src"), relativePath);
  }

  public testPath(relativePath = "") {
    return path.join(this.path("test"), relativePath);
  }

  public distPath(relativePath = "") {
    return path.join("dist", this.relativePath(relativePath));
  }

  public jestConfigPath() {
    return this.path("jest.config.ts");
  }

  public getTree() {
    return this.tree;
  }

  public getScope() {
    const { npmScope } = getWorkspaceLayout(this.tree);
    return npmScope;
  }

  public getName() {
    return this.names.kebabCase;
  }

  public getNames() {
    return this.names;
  }

  public getType() {
    return this.type;
  }

  private baseDir(relativePath = "") {
    const { appsDir, libsDir } = getWorkspaceLayout(this.tree);

    const isPackagesLayout = path.basename(libsDir) === "packages";

    let baseDir = libsDir;
    if (
      (isPackagesLayout && this.type === "e2e") ||
      (!isPackagesLayout && this.type === "application")
    ) {
      baseDir = appsDir;
    }

    return path.join(baseDir, relativePath);
  }
}
