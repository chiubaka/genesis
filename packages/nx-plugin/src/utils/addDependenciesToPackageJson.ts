import {
  addDependenciesToPackageJson as nxAddDependenciesToPackageJson,
  Tree,
} from "@nrwl/devkit";

import { getLatestPackageVersion } from "./getLatestPackageVersion";

type Dependencies = string[] | DependenciesWithVersions;
type DependenciesWithVersions = Record<string, string | undefined>;

export const addDependenciesToPackageJson = async (
  tree: Tree,
  dependencies: Dependencies,
  devDependencies: Dependencies,
) => {
  const dependenciesWithVersions = await calculateDependencyVersions(
    dependencies,
  );
  const devDependenciesWithVersions = await calculateDependencyVersions(
    devDependencies,
  );

  return nxAddDependenciesToPackageJson(
    tree,
    dependenciesWithVersions,
    devDependenciesWithVersions,
  );
};

const calculateDependencyVersions = async (
  dependencies: Dependencies,
): Promise<Record<string, string>> => {
  const dependenciesWithVersions: Record<string, string> = {};

  if (isDependencyArray(dependencies)) {
    for (const dependencyName of dependencies) {
      const version = await getLatestPackageVersion(dependencyName);
      // eslint-disable-next-line security/detect-object-injection
      dependenciesWithVersions[dependencyName] = version;
    }
  } else {
    for (const dependencyName in dependencies) {
      const version =
        // eslint-disable-next-line security/detect-object-injection
        dependencies[dependencyName] ||
        (await getLatestPackageVersion(dependencyName));
      // eslint-disable-next-line security/detect-object-injection
      dependenciesWithVersions[dependencyName] = version;
    }
  }

  return dependenciesWithVersions;
};

const isDependencyArray = (
  dependencies: Dependencies,
): dependencies is string[] => {
  return Array.isArray(dependencies);
};
