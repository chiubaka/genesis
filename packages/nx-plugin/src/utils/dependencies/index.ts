export * from "./addDependenciesToPackageJson";
export * from "./compatiblePackageVersions";
