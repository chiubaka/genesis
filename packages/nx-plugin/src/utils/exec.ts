import { exec as originalExec } from "node:child_process";
import { promisify } from "node:util";

export const exec = promisify(originalExec);
