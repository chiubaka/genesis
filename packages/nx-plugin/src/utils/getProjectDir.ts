import { getWorkspaceLayout, ProjectType, Tree } from "@nrwl/devkit";
import path from "node:path";

export const getProjectDir = (
  tree: Tree,
  projectName: string,
  projectType: ProjectType,
) => {
  const { appsDir, libsDir } = getWorkspaceLayout(tree);
  const baseDir = projectType === "application" ? appsDir : libsDir;

  return path.join(baseDir, projectName);
};
