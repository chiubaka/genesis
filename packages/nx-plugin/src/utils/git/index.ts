export * from "./getGitHubRepoInfo";
export * from "./parseRemoteUrl";
