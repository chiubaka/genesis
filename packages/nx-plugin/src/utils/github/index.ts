import { GitHubApiAdapter } from "./GitHubApiAdapter";

export {
  BranchProtection as GitHubBranchProtection,
  Label as GitHubLabel,
  Repo as GitHubRepo,
} from "./GitHubApiAdapter";

export const github = new GitHubApiAdapter();
