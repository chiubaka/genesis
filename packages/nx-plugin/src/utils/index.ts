export * from "./addDependenciesToPackageJson";
export * from "./exec";
export * from "./getProjectDir";
export * from "./git";
export * from "./github";
export * from "./Project";
export * from "./projectNameCasings";
export * from "./tasks";
