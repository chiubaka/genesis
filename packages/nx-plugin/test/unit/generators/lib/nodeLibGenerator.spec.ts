import { createTreeWithEmptyWorkspace } from "@chiubaka/nx-plugin-testing";
import { ProjectConfiguration, readJson, Tree } from "@nrwl/devkit";
import { PackageJson } from "nx/src/utils/package-json";

import {
  DockerComposeConfig,
  DockerComposeService,
  nodeLibGenerator,
  Project,
  readYaml,
  YarnConfig,
} from "../../../../src";
import { nodeProjectTestCases } from "../../../cases";

describe("nodeLibGenerator", () => {
  let tree: Tree;
  let project: Project;
  let e2eProject: Project;

  const getProject = () => {
    return project;
  };
  const getE2eProject = () => {
    return e2eProject;
  };

  beforeAll(async () => {
    tree = createTreeWithEmptyWorkspace();
    project = new Project(tree, "node-lib", "library");
    e2eProject = new Project(tree, "node-lib-e2e", "e2e");

    const projectScope = project.getScope();
    const projectName = project.getName();

    tree.write(
      "README.md",
      `[![codecov](https://codecov.io/gh/${projectScope}/${projectName}/branch/master/graph/badge.svg?token=foobar)](https://codecov.io/gh/${projectScope}/${projectName})`,
    );

    await nodeLibGenerator(tree, {
      name: "node-lib",
      skipE2e: false,
      localRegistry: "http://localhost:4873",
    });
  });

  nodeProjectTestCases(getProject, {
    projectJson: {
      targetNames: ["lint", "build", "test", "publish:local"],
    },
  });

  it("generates a sample unit test file in pascal case", () => {
    // eslint-disable-next-line security/detect-non-literal-fs-filename
    expect(tree.exists(project.testPath("unit/hello.spec.ts"))).toBe(true);
  });

  it("generates a sample file in pascal case", () => {
    // eslint-disable-next-line security/detect-non-literal-fs-filename
    expect(tree.exists(project.srcPath("hello.ts"))).toBe(true);
  });

  describe("package.json", () => {
    let packageJson: PackageJson;

    beforeAll(() => {
      packageJson = readJson<PackageJson>(tree, project.path("package.json"));
    });

    describe("scripts", () => {
      it("generates a deploy script", () => {
        expect(packageJson.scripts?.deploy).toBeTruthy();
      });

      it("generates a deploy:ci script", () => {
        expect(packageJson.scripts?.["deploy:ci"]).toBeTruthy();
      });
    });
  });

  describe("E2E project", () => {
    nodeProjectTestCases(getE2eProject, {
      projectJson: {
        targetNames: ["lint", "build", "e2e"],
      },
      repoName: "node-lib",
    });

    it("generates a main file", () => {
      // eslint-disable-next-line security/detect-non-literal-fs-filename
      expect(tree.exists(e2eProject.srcPath("main.ts"))).toBe(true);
    });

    it("generates a file that enhances exported members of the lib", () => {
      // eslint-disable-next-line security/detect-non-literal-fs-filename
      expect(tree.exists(e2eProject.srcPath("helloExtended.ts"))).toBe(true);
    });

    it("generates a yarn.lock file to keep this project separate from the parent workspace", () => {
      // eslint-disable-next-line security/detect-non-literal-fs-filename
      expect(tree.exists(e2eProject.path("yarn.lock"))).toBe(true);
    });

    it("generates a .yarnrc.yml file", () => {
      // eslint-disable-next-line security/detect-non-literal-fs-filename
      expect(tree.exists(e2eProject.path(".yarnrc.yml"))).toBe(true);
    });

    describe(".yarnrc.yml", () => {
      let yarnConfig: YarnConfig;

      beforeAll(() => {
        yarnConfig = readYaml<YarnConfig>(tree, e2eProject.path(".yarnrc.yml"));
      });

      it("uses the node-modules nodeLinker", () => {
        expect(yarnConfig.nodeLinker).toBe("node-modules");
      });

      it("uses the local registry to install packages", () => {
        expect(yarnConfig.npmRegistryServer).toBe("http://localhost:4873");
      });

      it("whitelists localhost for http", () => {
        expect(yarnConfig.unsafeHttpWhitelist).toEqual(["localhost"]);
      });
    });

    describe("project.json", () => {
      describe("e2e target", () => {
        it("depends on the library's build job", () => {
          const projectJson = readJson<ProjectConfiguration>(
            tree,
            e2eProject.path("project.json"),
          );

          expect(projectJson.implicitDependencies).toContain("node-lib");
          expect(projectJson.targets?.e2e.dependsOn).toContainEqual({
            target: "publish:local",
            projects: "dependencies",
          });
        });
      });
    });
  });

  describe("workspace", () => {
    describe("package.json", () => {
      it.todo("Adds a start:verdaccio script");
    });

    describe("README.md", () => {
      it.todo(
        "Updates the README with instructions about how to run E2E tests",
      );
    });

    describe("docker-compose", () => {
      it("generates a docker-compose.yml file for the workspace", () => {
        expect(tree.exists("docker-compose.yml")).toBe(true);
      });

      describe("docker-compose.yml", () => {
        let dockerComposeConfig: DockerComposeConfig;

        beforeAll(() => {
          dockerComposeConfig = readYaml<DockerComposeConfig>(
            tree,
            "docker-compose.yml",
          );
        });

        describe("registry", () => {
          let registryConfig: DockerComposeService;

          beforeAll(() => {
            registryConfig = dockerComposeConfig.services.registry;
          });

          it("defines a registry service", () => {
            expect(registryConfig).toBeDefined();
          });

          it("creates a verdaccio service", () => {
            expect(registryConfig.image).toBe("verdaccio/verdaccio");
          });

          it("runs verdaccio on port 4873", () => {
            expect(registryConfig.ports).toContain("4873:4873");
          });

          it("uses the generated verdaccio config in the docker container", () => {
            expect(registryConfig.volumes).toContain(
              "./verdaccio:/verdaccio/conf",
            );
          });
        });
      });

      describe("verdaccio", () => {
        it("generates a basic verdaccio config", () => {
          expect(tree.exists("verdaccio/config.yaml")).toBe(true);
        });

        describe("config.yaml", () => {
          it.todo(
            "ensures that the access to the generated package is not proxied to NPM",
          );
        });
      });
    });
  });
});
