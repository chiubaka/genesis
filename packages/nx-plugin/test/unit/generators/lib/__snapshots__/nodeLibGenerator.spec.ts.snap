// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodeLibGenerator .eslintrc.json matches snapshot 1`] = `
Object {
  "extends": Array [
    "../../.eslintrc.json",
  ],
  "ignorePatterns": Array [
    "node_modules",
  ],
}
`;

exports[`nodeLibGenerator E2E project .eslintrc.json matches snapshot 1`] = `
Object {
  "extends": Array [
    "../../.eslintrc.json",
  ],
  "ignorePatterns": Array [
    "node_modules",
  ],
}
`;

exports[`nodeLibGenerator E2E project project.json matches snapshot 1`] = `
Object {
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "implicitDependencies": Array [
    "node-lib",
  ],
  "projectType": "application",
  "sourceRoot": "e2e/node-lib-e2e/src",
  "tags": Array [],
  "targets": Object {
    "build": Object {
      "configurations": Object {
        "production": Object {
          "extractLicenses": true,
          "fileReplacements": Array [
            Object {
              "replace": "e2e/node-lib-e2e/src/environments/environment.ts",
              "with": "e2e/node-lib-e2e/src/environments/environment.prod.ts",
            },
          ],
          "inspect": false,
          "optimization": true,
        },
      },
      "executor": "@nrwl/node:webpack",
      "options": Object {
        "assets": Array [
          "e2e/node-lib-e2e/src/assets",
        ],
        "main": "e2e/node-lib-e2e/src/main.ts",
        "outputPath": "dist/e2e/node-lib-e2e",
        "tsConfig": "e2e/node-lib-e2e/tsconfig.app.json",
      },
      "outputs": Array [
        "{options.outputPath}",
      ],
    },
    "e2e": Object {
      "dependsOn": Array [
        Object {
          "projects": "self",
          "target": "install",
        },
      ],
      "executor": "@nrwl/jest:jest",
      "options": Object {
        "jestConfig": "e2e/node-lib-e2e/jest.config.ts",
        "passWithNoTests": true,
      },
      "outputs": Array [
        "reports/coverage/e2e/node-lib-e2e",
        "reports/junit/node-lib-e2e.xml",
      ],
    },
    "install": Object {
      "dependsOn": Array [
        Object {
          "projects": "dependencies",
          "target": "build",
        },
      ],
      "executor": "@nrwl/workspace:run-commands",
      "options": Object {
        "commands": Array [
          "yarn cache clean --all",
          "yarn install",
        ],
        "cwd": "e2e/node-lib-e2e",
        "parallel": false,
      },
    },
    "lint": Object {
      "executor": "@nrwl/linter:eslint",
      "options": Object {
        "lintFilePatterns": Array [
          "e2e/node-lib-e2e",
        ],
      },
      "outputs": Array [
        "{options.outputFile}",
      ],
    },
  },
}
`;

exports[`nodeLibGenerator E2E project tsconfig.app.json or tsconfig.lib.json matches snapshot 1`] = `
Object {
  "compilerOptions": Object {
    "declaration": true,
    "types": Array [
      "node",
    ],
  },
  "exclude": Array [
    "jest.config.ts",
    "**/*.spec.ts",
    "**/*.test.ts",
    "test",
  ],
  "extends": "./tsconfig.json",
  "include": Array [
    "**/*.ts",
  ],
}
`;

exports[`nodeLibGenerator E2E project tsconfig.json matches snapshot 1`] = `
Object {
  "compilerOptions": Object {
    "lib": Array [
      "es2022",
    ],
    "module": "commonjs",
    "outDir": "../../dist/out-tsc",
    "target": "es2022",
  },
  "extends": "../../tsconfig.base.json",
  "files": Array [],
  "include": Array [],
  "references": Array [
    Object {
      "path": "./tsconfig.lib.json",
    },
    Object {
      "path": "./tsconfig.spec.json",
    },
  ],
}
`;

exports[`nodeLibGenerator E2E project tsconfig.spec.json matches snapshot 1`] = `
Object {
  "compilerOptions": Object {
    "types": Array [
      "jest",
      "node",
    ],
  },
  "extends": "./tsconfig.json",
  "include": Array [
    "jest.config.ts",
    "**/*.test.ts",
    "**/*.spec.ts",
    "**/*.d.ts",
    "test",
  ],
}
`;

exports[`nodeLibGenerator project.json matches snapshot 1`] = `
Object {
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "library",
  "sourceRoot": "packages/node-lib/src",
  "tags": Array [],
  "targets": Object {
    "build": Object {
      "executor": "@nrwl/js:tsc",
      "options": Object {
        "assets": Array [
          "packages/node-lib/*.md",
        ],
        "main": "packages/node-lib/src/index.ts",
        "outputPath": "dist/packages/node-lib",
        "packageJson": "packages/node-lib/package.json",
        "tsConfig": "packages/node-lib/tsconfig.lib.json",
      },
      "outputs": Array [
        "{options.outputPath}",
      ],
    },
    "lint": Object {
      "executor": "@nrwl/linter:eslint",
      "options": Object {
        "lintFilePatterns": Array [
          "packages/node-lib",
        ],
      },
      "outputs": Array [
        "{options.outputFile}",
      ],
    },
    "test": Object {
      "executor": "@nrwl/jest:jest",
      "options": Object {
        "jestConfig": "packages/node-lib/jest.config.ts",
        "passWithNoTests": true,
      },
      "outputs": Array [
        "reports/coverage/packages/node-lib",
        "reports/junit/node-lib.xml",
      ],
    },
  },
}
`;

exports[`nodeLibGenerator tsconfig.app.json or tsconfig.lib.json matches snapshot 1`] = `
Object {
  "compilerOptions": Object {
    "declaration": true,
    "types": Array [
      "node",
    ],
  },
  "exclude": Array [
    "jest.config.ts",
    "**/*.spec.ts",
    "**/*.test.ts",
    "test",
  ],
  "extends": "./tsconfig.json",
  "include": Array [
    "**/*.ts",
  ],
}
`;

exports[`nodeLibGenerator tsconfig.json matches snapshot 1`] = `
Object {
  "compilerOptions": Object {
    "lib": Array [
      "es2022",
    ],
    "module": "commonjs",
    "outDir": "../../dist/out-tsc",
    "target": "es2022",
  },
  "extends": "../../tsconfig.base.json",
  "files": Array [],
  "include": Array [],
  "references": Array [
    Object {
      "path": "./tsconfig.lib.json",
    },
    Object {
      "path": "./tsconfig.spec.json",
    },
  ],
}
`;

exports[`nodeLibGenerator tsconfig.spec.json matches snapshot 1`] = `
Object {
  "compilerOptions": Object {
    "types": Array [
      "jest",
      "node",
    ],
  },
  "extends": "./tsconfig.json",
  "include": Array [
    "jest.config.ts",
    "**/*.test.ts",
    "**/*.spec.ts",
    "**/*.d.ts",
    "test",
  ],
}
`;
